packages/sql/src/visit/walk.js

```diff
@@ -0,0 +1,30 @@
+import { isNode } from '../ast/node.js';
+import { recurse } from './recurse.js';
+
+/**
+ * Perform a traversal of a SQL expression AST.
+ * @param {import('../ast/node.js').SQLNode} node Root node for AST traversal.
+ * @param {import('../types.js').VisitorCallback} visit Visitor callback function.
+ * @return {import('../types.js').VisitorResult}
+ */
+export function walk(node, visit) {
+  if (!isNode(node)) return;
+  const result = visit(node);
+  if (result) return result;
+
+  const props = recurse[node.type];
+  const n = props?.length ?? 0;
+  for (let i = 0; i < n; ++i) {
+    const value = node[props[i]];
+    if (Array.isArray(value)) {
+      const m = value.length;
+      for (let j = 0; j < m; ++j) {
```

Comment:
Could be a for...of

Reply:
It could, but I wanted to keep the loops tight in cases of larger queries/expressions to walk. Using iterators adds non-trivial overhead. Maybe wouldn't really matter in this case, but I prefer to default to the more performant case for internals like this.